src/relations/database.py

```diff
@@ -0,0 +1,124 @@
+
+        self.framework.observe(
+            self.charm.on[DB_RELATION_NAME].relation_broken, self._on_database_broken
+        )
+
+    def _get_or_set_password(self, relation) -> str:
+        """Retrieve password from cache or generate a new one.
+
+        Args:
+            relation (str): The relation for each the password is cached.
+
+        Returns:
+            str: The password.
+        """
+        if password := relation.data[self.charm.app].get("password"):
+            return password
+        password = generate_random_password(PASSWORD_LENGTH)
+        relation.data[self.charm.app]["password"] = password
+        return password
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent):
+        """Handle the `database-requested` event."""
+        if not self.charm.unit.is_leader():
+            return
+        # check if cluster is ready and if not, defer
+        if not self.charm.cluster_initialized:
+            logger.debug("Waiting cluster to be initialized")
+            event.defer()
+            return
+
+        # get base relation data
+        relation_id = event.relation.id
+        db_name = event.database
+        # user name is derived from the relation id
+        db_user = f"relation-{relation_id}"
+        db_pass = self._get_or_set_password(event.relation)
+        db_version = self.charm._mysql.get_mysql_version()
+
+        remote_app = event.app.name
+
+        try:
+            primary_endpoint = self.charm._mysql.get_cluster_primary_address()
+            self.database.set_credentials(relation_id, db_user, db_pass)
+            self.database.set_endpoints(relation_id, primary_endpoint)
+            self.database.set_version(relation_id, db_version)
+            # get read only endpoints by removing primary from all members
+            read_only_endpoints = sorted(
+                self.charm._mysql.get_cluster_members_addresses()
+                - {
+                    primary_endpoint,
+                }
+            )
+
+            self.database.set_read_only_endpoints(relation_id, ",".join(read_only_endpoints))
+            # TODO:
+            # add setup of tls, tls_ca and status
+            # add extra roles parsing from relation data
+            self.charm._mysql.create_application_database_and_scoped_user(
+                db_name, db_user, db_pass, "%", remote_app
+            )
+
+            logger.info(f"Created user for app {remote_app}")
+        except MySQLCreateApplicationDatabaseAndScopedUserError:
+            logger.error(f"Failed to create scoped user for app {remote_app}")
+            self.charm.unit.status = BlockedStatus("Failed to create scoped user")
+            return
+        except MySQLClientError as e:
+            logger.exception("Failed to retrieve cluster info", exc_info=e)
+            self.charm.unit.status = BlockedStatus("Failed to retrieve cluster info")
+            return
```

Comment:
do you need these `return`s? they're at the end of the branch for the method. same with line 124

Reply:
not really